scripts/verify-install.sh

```diff
@@ -119,23 +126,37 @@ if [[ -f /shasums.txt ]]; then
   fi
 fi
 
-# Import all the signing keys.
+# A trick to get the absolute directory where this script is located, no matter
+# how or from where it was called. We'll need it to locate the key files which
+# are located relative to this script.
+DIR="$(cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd)"
+
+# Import all the signing keys. We'll create a key ring for each user and use
+# that exact key ring when verifying a user's signature. That way we can make
+# sure one user cannot just upload multiple signatures to reach the 5/7 required
+# sigs.
+TEMP_DIR=$(mktemp -d /tmp/lnd-sig-verification-XXXXXX)
 for key in "${KEYS[@]}"; do
   KEY_ID=$(echo $key | cut -d' ' -f1)
-  IMPORT_URL=$(echo $key | cut -d' ' -f2)
-  echo "Downloading and importing key $KEY_ID from $IMPORT_URL"
-  curl -L -s $IMPORT_URL | gpg --import
+  IMPORT_FILE=$(echo $key | cut -d' ' -f2)
+  KEY_FILE="$DIR/$IMPORT_FILE"
+  BASENAME=$(basename "$KEY_FILE")
+  USERNAME=${BASENAME%.*}
+  KEYRING="$TEMP_DIR/$USERNAME.pgp"
+  echo "Importing key for user $USERNAME from $key"
+  gpg --no-default-keyring --keyring "$KEYRING" --import < "$KEY_FILE"
 
   # Make sure we actually imported the correct key.
-  if ! gpg --list-key "$KEY_ID"; then
-    echo "ERROR: Imported key from $IMPORT_URL doesn't match ID $KEY_ID."
+  if ! gpg --no-default-keyring --keyring "$KEYRING" --list-key "$KEY_ID"; then
```

Comment:
I believe it's possible to have two keys in a single GPG file so if we assume the files can get compromised then this check would fail to detect that. If we assume the files aren't compromised this check is useless.

Reply:
This is a sanity check that the key file wasn't updated without also changing the ID in this file. We should perhaps run that check during CI and not during verification itself.

Reply:
My point is that an attacker can trivially cause this script to import multiple keys - including an attacker-controlled one. (See below.)

I now understand the sanity check but I wonder what the assumptions made here even are? What's the threat model? Is the contents of git repository assumed to be trusted? If yes, why not just commit the single fingerprint of a binary? If no, then should the user be responsible for verifying all keys *manually or with their script*? (If yes, can I contribute a well-written script somewhere?)

I simply assumed that the only trusted components are the list of fingerprints and the logic of the script itself. The approach to treat everything else as malicious until proven otherwise is quite common, but you may have a different view.

Until this is cleared out, we will probably have unproductive back-and-forth which I believe we both would prefer to avoid. :)

Here's a simple demonstration of the problem described above:

```bash
gpg --quick-generate-key Alice
gpg --quick-generate-key Bob
gpg --export Alice > combined.gpg
gpg --export Bob >> combined.gpg
GNUPGHOME=/tmp/gpg gpg --import < combined.gpg
```

Contrary to what you might think, the last command imports both keys:

```
gpg: key 6304D22222D6D832: public key "Alice" imported
gpg: key A8AE0C045E5A0C59: public key "Bob" imported
gpg: Total number processed: 2
gpg:               imported: 2
```

Output from `gpg --list-keys`:

```
-----------------------------
pub   rsa3072 2021-03-11 [SC] [expires: 2023-03-11]
      71E9E15B66A54103117F0D4A6304D22222D6D832
uid           [ultimate] Alice
sub   rsa3072 2021-03-11 [E]

pub   rsa3072 2021-03-11 [SC] [expires: 2023-03-11]
      9F59A351932839AD8E48E4B6A8AE0C045E5A0C59
uid           [ultimate] Bob
sub   rsa3072 2021-03-11 [E]
```

Therefore if you blindly import keys and then assume that all keys in the keyring are trusted, you're vulnerable to attacks. My script avoided exactly this problem by using a temporary keyring and then exporting *the specific key only* from this keyring into a final keyring.

For the record my project [`sqck`](https://github.com/Kixunil/sqck) happens to avoid this problem and a bunch of others.

Reply:
Thank you for your detailed explanation. I think we are definitely operating with different threat models in mind. Or at least when this script was first created, this was already quite a big step forward compared to the single (or some times double) signatures you had to manually verify and couldn't reproduce from the code yourself.

But now that we have _something_, we can start improving upon it, so thank you for your input!
This is how I would describe our (current) threat model (other people on the team might disagree):
 - The content of the git repository is seen as sufficiently trusted for now. We commit the keys to the git repository to make it more explicit and publicly visible (through a PR) if a key changes. And that way we'd see if there were two keys in a file. Of course this is completely useless if you must assume, during runtime of `verify-install`, that the key files might have been modified by an attacker and don't correspond to what's in the git repo anymore. I'm not sure what can be done in that case. Relying on key servers to fetch the keys seems very brittle these days. But I think for now, we could add an additional check to the script that an imported key file contains _only_ the key we want.
 - The content of the GitHub _release_ page is not trusted. This script focuses on protecting users from maliciously uploaded release binaries. For example if the GitHub account of a privileged user is hijacked, we want to make it impossible for that single user to be able to replace the release binaries and upload fake signatures. Uploading files to a release is quite hidden, but if someone started pushing changes to the PGP keys to try to "legitimize" their uploaded fake signatures, that would hopefully be detected immediately.

Your question "what script should I use then? this one or my own?": I think people also shouldn't blindly trust in a provided script. The goal is, as I said, to have a reasonably easy to use way to be _reasonably_ certain that the _installed_ binaries are coming from an "official" release, both in a Docker and non-docker environment.
Everything that exceeds this "basic" use case must probably be tailored to each user's individual threat model anyway and perhaps cannot be stuffed into a "one-that-fits-all" script.
Or perhaps this is a naive approach and we should only have the most paranoid script on our repo in the first place. But my personal opinion is: If the verification is too involved (manual process, keyserver connectivity issues, unclear usage, requirement for installed tools), people won't use it. If it's easy to use, more users actually verify their installation, but perhaps not with a script that could defend against state level actors.

Long story short: I think we should make this script as secure as we can under the assumption that the git content is not modified/malicious and that we want to verify the _installed_ binary. Everything that is more involved should either go into a separate script (for example for verifying a downloaded tarball) or be optional features activated through flags (for example a `--use-keyservers` flag).

With the above said, do you think we can proceed with the approach in this PR, if I add the following two changes? Or does this still not add any increased security to the use case you envision?
 - Make sure only one key is imported from a key file (in addition to making sure it's the one we expect).
 - Add an option `--use-keyservers` to cross reference/download the keys from external keyservers

Reply:
## Short answer

Great, I think we can move forward now. Making sure there's exactly one key imported sounds good, assuming it's bug-free (don't count sub-keys etc). I'm not sure why you suggest use of keyservers, I didn't intend to suggest they improve anything. While I personally didn't have problems with the Ubuntu keyserver, I don't have any reason to recommend it over having the keys in git either.

## Bunch of opinions on the topic of security and some general suggestions

There are various groups of people who will have various approaches to security:

* Paranoid (like me :D), who review the verification script or write their own and also cross-verify the fingerprints/pubkeys in person or from different sources (GitHub, Keybase, Twitter...) These people can not be attacked by anyone except the developers.
* TOFU users - they would assume the script and the keys they got initially are valid without reviewing them and continue to use them in followup releases. Thus an attacker would have to compromise the initial release to succeed. This shortens window of opportunity greatly. While I myself try to avoid this, I still consider it very useful, probably great for most people.
* People who do quick sanity checks - they check hashes or something trivial. This is not strictly secure but attackers are sometimes lazy or stupid, so mere checking of hash can save their ass.
* People who don't verify anything at all for various reasons.

Out of all these groups, only the paranoid ones will actually read the verification script. Having intentionally weak script because it's easier to read doesn't seem to make a lot of sense because people who give it really proper look can see the flaws and will avoid using such script and likely report it as bug. To them the script is useless so if they wrote their own instead of spending time reviewing they are better-off. If the script is complicated but with good explanation of those complications, the paranoid people will see why it makes sense.

The amount of internal complication doesn't matter to the remaining groups because they don't read the script. Thus making the script better inside has only benefits and zero downsides for these people.

You bring a very good point about having simple interface. `verify.sh VERSION something.tar.gz` seems simple enough. Anything simpler is probably done by some automated software management invisible to the user and the user just TOFUs the manager. Well, this is one of the foundational ideas behind my Deb repository. :)

What I want to say is that making the internal logic of the script maximally secure seems to be the rational choice.

> I think people also shouldn't blindly trust in a provided script.

That'd be a really secure world but sadly I don't think everyone including my mother will review verification scripts. Manual verification can also be problematic. People get tired and make mistakes. Sometimes really bad mistakes. The often underappreciated value of automation is reliability. So I believe getting as many people to *at least* use TOFU script, perhaps mixed with reputation/WOT can have much bigger effect on security of wide population. Especially if they TOFU something well-written. Placing a well-written script in a reputable repository like this one is the best way to ensure people use it.

Regarding having multiple scripts, I think there are serious downsides that should be considered. A huge strength of code reuse is the ability to join review capability of multiple developers who complement each-other and thus find more bugs together. Thus a bit larger but more reviewed piece of code can easily be better than everyone writing their own script.

Feature bloat can be a problem in general but I don't think it's likely to apply in this case. Verification has in general simple inputs: a program, signature and a set of public keys. There can be variations on how exactly they are specified but there isn't ton of them. I can't imagine tens of people asking for new features to be implemented in verification script.

Frankly, if the idea is to keep it simple, I'd suggest to support verification *before* installation only. Most secure systems verify signatures before touching the signed data. It can avoid various issues. An interesting example is the ability to break unauthenticated AES in some scenarios. This is not something that bothers me too much and I can accept your decision to verify the installed binaries instead.

If I managed to convince you that trying to push people from non-verification towards TOFU and from manual verification to automated makes sense, then maybe you'll also agree to improving verification documentation. I have some ideas to make it shorter and safer by explicitly suggesting non-technical users to TOFU properly and suggesting technical users to review the script and publish the hash at various places.

Reply:
> Having intentionally weak script because it's easier to read doesn't seem to make a lot of sense

I didn't mean that the script has to be simple to _read_. Only simple to use, meaning: limit the number of possible options/flag combinations, number of runtime requirements (keys on disk vs. dependency on keyservers) or software installed.
The script itself should be written as securely as it can be, as long as its usage is easy enough for most users to actually run it.

> I'm not sure why you suggest use of keyservers

Using some sort of key server (PGP or Keybase or whatever) is the only way to get away with having just one script file. But for now I think having the keys in git so we can track their changes more explicitly is probably preferred (by us) over just having one script. Maybe we can add other key download options in a follow-up PR.

Reply:
Sure, that makes sense. One should only need version and file to verify, don't see the need for anything else.

Ah, I see you used "keyservers" in broader sense.

>  having just one script file

I think the goal should be having *to verify* one script file. If you see fingerprints in the script file and the script behaves correctly then you only need to verify that the fingerprints belong to the people you trust. Verifying fingerprints is easier than whole keys.